### PR TITLE
Skip image requests that previously 404'd

### LINKED ISSUE #2714: Add a 404 "skip list" to WPNetworkImageView
WPNetworkImageView should be smart enough to skip images that have already returned a 404. This is especially important for avatars & blavatars, since we often request ones that don't exist and continue to request them even though we should already know they don't exist.
```
05-17 17:37:16.316    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/ecbf88040c5ac307d2a0ccfa65d5b7cb?s=168&d=404
05-17 17:37:16.409    4802-4887/org.wordpress.android E/WordPress﹕ [15028] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/c9deccf60483e4b42441ec2465b7280f?s=168&d=404
05-17 17:37:21.347    4802-4887/org.wordpress.android E/WordPress﹕ [15028] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/d26e4ac1c55e9a1ce645544d9db54ffc?s=168&d=404
05-17 17:37:26.425    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/15d90339ea749223c5e72f95a9800a7d?s=168&d=404
05-17 17:37:49.741    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/29b80dcd9467a130200b004ac9b5fe52?s=168&d=404
05-17 17:37:50.914    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/74897ad976de8dfaa643b9f2a209b66d?s=168&d=404
05-17 17:37:51.437    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/5d5c6a88e20efec45774eb56ddc6b8ca?s=168&d=404
05-17 17:37:51.581    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/fe86f328eeabc8af23378fe0ed7c6f0a?s=168&d=404
05-17 17:37:56.011    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4bd22a5105962cd99aabff3c36627aa3?s=168&d=404
05-17 17:37:56.069    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/29b80dcd9467a130200b004ac9b5fe52?s=168&d=404
05-17 17:37:56.157    4802-4887/org.wordpress.android E/WordPress﹕ [15028] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/c12c8eb573921fb31af7f49c3fe4b23f?s=168&d=404
05-17 17:37:59.022    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/29b80dcd9467a130200b004ac9b5fe52?s=168&d=404
05-17 17:37:59.523    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4cac5630c0452dbbe6edc33fb80e1829?s=168&d=404
05-17 17:38:00.023    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/cb9ecf3780bc637f9172fc09526d3d9e?s=168&d=404
05-17 17:38:00.107    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/e53f35ceaebfa7ba670f8d2e7d66b4d4?s=168&d=404
05-17 17:38:01.668    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/cb9ecf3780bc637f9172fc09526d3d9e?s=168&d=404
05-17 17:38:04.560    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4cac5630c0452dbbe6edc33fb80e1829?s=168&d=404
05-17 17:38:04.930    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://2.gravatar.com/avatar/29b80dcd9467a130200b004ac9b5fe52?s=168&d=404
05-17 17:38:05.118    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/c12c8eb573921fb31af7f49c3fe4b23f?s=168&d=404
05-17 17:38:05.607    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4bd22a5105962cd99aabff3c36627aa3?s=168&d=404
05-17 17:38:07.355    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/c12c8eb573921fb31af7f49c3fe4b23f?s=168&d=404
05-17 17:38:07.460    4802-4887/org.wordpress.android E/WordPress﹕ [15028] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4bd22a5105962cd99aabff3c36627aa3?s=168&d=404
05-17 17:38:09.497    4802-4888/org.wordpress.android E/WordPress﹕ [15029] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4bd22a5105962cd99aabff3c36627aa3?s=168&d=404
05-17 17:38:10.520    4802-4886/org.wordpress.android E/WordPress﹕ [15027] BasicNetwork.performRequest: Unexpected response code 404 for https://1.gravatar.com/avatar/4bd22a5105962cd99aabff3c36627aa3?s=168&d=404
05-17 17:38:10.537    4802-4889/org.wordpress.android E/WordPress﹕ [15030] BasicNetwork.performRequest: Unexpected response code 404 for https://0.gravatar.com/avatar/c12c8eb573921fb31af7f49c3fe4b23f?s=168&d=404
```

### PR DESCRIPTION
**Pull request by nbradbury**
Fix #2714 - WPNetworkImageView now keeps track of image requests which return a 404 so it can skip requesting them in the future (during the same session).

### CODE REVIEW

**Comment by daniloercoli** [comment]
:clap:

File: WordPress/src/main/java/org/wordpress/android/widgets/WPNetworkImageView.java [hunk 1]
```diff
@@ -183,6 +187,13 @@ private void loadImageIfNecessary(final boolean isInLayoutPass) {
             }
         }
 
+        // skip this URL if a previous request for it returned a 404
+        if (mUrlSkipList.contains(mUrl)) {
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/widgets/WPNetworkImageView.java]: I wonder If we should call `add` and `contains` in a synchronized way.
I see in the documentation of `Hashset` <blockquote>Note that this implementation is not synchronized. If multiple threads access a hash set concurrently, and at least one of the threads modifies the set, it must be synchronized externally.</blockquote>

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/widgets/WPNetworkImageView.java]: > I wonder If we should call add and contains in a synchronized way.

I'm pretty sure `mUrlSkipList` is only accessed from the main thread, so synchronization isn't necessary.

**Comment by daniloercoli** [comment]
:shipit:

**Pull request merged**